Use vectors in the Soft Limiter and upstream audio buffers [1/4]

kcgen opened the pull request:
The primary motivation is that vectors are moveable, having dynamically allocated content. This makes for very efficient (almost no-op) passing of buffers from the soft-limiter into an audio queue, and from there back out to the realtime play buffer (all of which are now vectors).

This is an all or nothing change: once made in the limiter, it requires all up-stream users to change as well (GUS, FluidSynth, MT-32, and the unit tests).

The limiter is no longer a template class (which was previously required to size the std::arrays), so this is a welcome simplification. Roughly 300 lines-of-change in this PR are due to moving the soft limiter's code from header to .cpp.

Because vectors can't be statically sized like arrays, the small bits of new code take care of this initial sizing as well as `assert`s  to sanity-check the vectors' sizes.

Effort is also made to `reserve()` the vector capacity when possible, as in many cases the maximum number of frames is known at compile time, as those values are stored in `constexpr`s. This allows the compiler to perform the most efficient allocation as possible at compile time.

kcgen commented:
**Testing**:

- [X] Linux x86-64
- [ ] Linux ARM
- [X] Windows VM
- [  ] Windows hardware
- [  ] macOS

**Benchmarks**

Up coming. 
We should see equal (or slightly better) performance due to improved `move` behavior.

kcgen commented:
Finished up testing. Here are a couple playable recordings from macOS:
 - :movie_camera:  [std::vectors + FluidSynth]( https://drive.google.com/file/d/1laROnN6Be-uj52TQQLj_1q8J4ALGQTWW/view?usp=sharing)
 - :movie_camera:  [std::vectors + MT-32](https://drive.google.com/file/d/1ggAoESUgAacOcFUz8UEu7WXqZrBa1kGo/view?usp=sharing)

kcgen commented:
Benchmarked MT-32 playback using `std::array` (baseline) versus `std::vector` (PR).
 - Both builds were managed by the RW queue
 - Debug build
 - CPU set to 800 MHz to magnify performance differences
 - Game: Space Quest 3 running the dynamic core at 2000 cycles
 - Benchmark command:  `time dosbox sq3.bat & sleep 60 && killall dosbox`

Baseline (5 runs):
 - Wall times: 60.196,  60.273, 60.346, 60.189, 60.364
 - User times: 25.345, 25.058, 27.968, 29.730, 30.562
 - Sys times: 2.687, 2.155, 2.315, 3.088, 1.547
 - (Sum(User) + Sum(Sys)) / Sum(wall) = 49.92%

PR (5 runs):
 - Wall times: 60.383, 60.483, 60.227, 60.296, 60.155
 - User times: 27.968, 27.530, 26.664, 24.736, 25.179
 - Sys times: 2.316, 2.789, 2.472, 2.119, 2.179
 - (Sum(User) + Sum(Sys)) / Sum(wall) = 47.74%

dreamer commented on src/hardware/gus.cpp:
```diff
@@ -600,6 +601,9 @@ Gus::Gus(uint16_t port, uint8_t dma, uint8_t irq, const std::string &ultradir)
 	}
 
+	// Size the accumulating write buffer
+	accumulator.resize(BUFFER_SAMPLES);
```
I am still as confused about this `accumulator` name as I was since the very beginning. Accumulator is a single value, not a collection of values.

Aside of that, this resize seems **very suspicious**. Shouldn't it be a `reserve`? 

TODO for me to check later in review: what is this accumulator anyway? Does reallocating some vector here break other code?

kcgen replied:
Good point; it's an array (or now vector) of accumulators.  I'll try to rename it to be clearer.

If you look at the callback, it asks for ~1ms of audio.

When we enter the callback, we rendering one voice at a time:

``` c++
	auto v = voices.begin();
	const auto v_end = v + active_voices;
	while (v < v_end && *v) {
		v->get()->GenerateSamples(accumulator, ram, vol_scalars,
		                          pan_scalars, frames, dac_enabled);
		++v;
	}
```

`GenerateSamples` renders the audio for a single voice - say, a trumpet, over the requested 1ms worth of audio frames.

Notice we pass the `accumulator` array into `GenerateSamples(..)`. 

If we drill into GenerateSamples, you'll see the accumulator is now called 'stream' (because relative to this function, it's no longer an accumulator).  The stream values are `v` (unfortunately named for this example..) 

 But anyway, we see the trumpet's samples steadily are added to the accumulator values (the `*v ++ =`). It's not just an assignment.

``` c++
	// Add the samples to the stream, angled in L-R space
	while (v < last_v) {
		float sample = GetSample(ram);
		sample *= PopVolScalar(vol_scalars);
		if (dac_enabled) {
			*v++ += sample * pan_scalar.left;
			*v++ += sample * pan_scalar.right;
		}
	}
```

So now we've got accumulator holding 1 ms worth of trumpet.

We then step into next voice, say a drum, and literally add those samples onto the prior trumpet samples. 

That repeats for all active voices (up to 32 voices) until each accumulator [] position in the array has the sum of every voice :-)

kcgen replied:
Regarding renaming a couple of these poorly named variables, I placed this in a separate commit: 4c040f7c5

kcgen replied:
> Resize seems very suspicious. Shouldn't it be a reserve?

We only need to allocate the rendering buffer once (So I prefer to do that here in the GUS's constructor - one and for all).  After that, we reuse that array over and over, and never need to grow or shrink it.  

In general, for audio code you never want to do dymanic memory operations.  Great article here - it's all common sense, but it's nice and quick reading from someone who writes audio code for live concerts (one glitch, and your code will never be used again!). - http://atastypixel.com/blog/four-common-mistakes-in-audio-development/

`Reserve` does allocate the memory, but it doesn't let me use the elements until I  `push_back(elem)` (which I'm not going to do) or `.resize(elems)`.  So I'd rather just resize() straighy away.

Also, I resize by a `constexpr` value - so this allocation might even be optimized at compile-time.

kcgen replied:
> Aside of that, this resize seems very suspicious. Shouldn't it be a reserve?

Yes - much better now:

``` c++
Gus::Gus(uint16_t port, uint8_t dma, uint8_t irq, const std::string &ultradir)
        : render_buffer(BUFFER_FRAMES * 2), // 2 samples/frame (left & right)
```

And gets rid of the need for BUFFER_SAMPLES now!

dreamer commented on src/hardware/soft_limiter.cpp:
```diff
@@ -0,0 +1,299 @@
+          max_samples(max_frames * 2)
+{}
+
+// Applies the Soft Limiter to the given input sequence and returns the results
+// as a reference to a std::vector of 16-bit ints the same length as the input.
+typename SoftLimiter::out_t SoftLimiter::Apply(const in_t &in,
```
As a general rule: if there's a need to write non-trivial amount of documentation to explain how to use the function/method… it means the interface is not intuitive and should probably be adjusted.

kcgen replied:
That's a good general rule (and I agree with it). 

This is just a single sentence saying that the limiter will be applied to your input array and returned as an integer array.  I could simplify it further (to that wording), if it's a bit too verbose right now?

dreamer commented on src/hardware/soft_limiter.cpp:
Perhaps this should be a static function instead of "helper method"? Maybe I miss something, but I don't see a single member of the class being used here.

kcgen replied:
Yup - Will do.

kcgen replied:
Yup - good catch!

dreamer commented on src/midi/midi_fluidsynth.cpp:
```diff
@@ -178,6 +179,12 @@ static std::string find_sf_file(const std::string &name)
+MidiHandlerFluidsynth::MidiHandlerFluidsynth()
+        : soft_limiter("FSYNTH", prescale_level, expected_max_frames)
+{
+	stream.resize(expected_max_frames * 2); // L & R channels
```
This is suspicious.

kcgen replied:
(Same reason as the one-time resize in the SoftLimiter)

dreamer replied:
I commented on this in new round of review - this means this vector's constructor should be called with `size_t` parameter in this class initializer list.

kcgen replied:
Absolutely; thanks @dreamer.
There were some stack-exchange answers that discourage using vector's constructor to do most operations (default values, default size, etc..) and favour the explicit calls just for readability.  But I actually prefer moving this into the initializer list, and the default-size is on-par with std::array's default size template arg too.

dreamer commented on src/midi/midi_mt32.h:
```diff
@@ -45,16 +45,16 @@ class MidiHandler_mt32 final : public MidiHandler {
-	using render_buffer_t = std::array<float, SAMPLES_PER_BUFFER>;
-	using limited_buffer_t = std::array<int16_t, SAMPLES_PER_BUFFER>;
-	using ring_t = moodycamel::BlockingReaderWriterCircularBuffer<limited_buffer_t>;
+	using render_buffer_t = std::vector<float>;
+	using play_buffer_t = std::vector<int16_t>;
```
I think we can stop using these two type aliases now - simple `std::vector<type>` will be easier to read and understand.

kcgen replied:
Gone! Looking tidy :-)

dreamer commented on src/midi/midi_mt32.h:
```diff
@@ -45,16 +45,16 @@ class MidiHandler_mt32 final : public MidiHandler {
 	static constexpr int FRAMES_PER_BUFFER = 1024; // synth granularity
 	static constexpr int SAMPLES_PER_BUFFER = FRAMES_PER_BUFFER * 2; // L & R
```
These static constexpr class members can now be hidden in cpp file, they no longer need to be exposed in header file, I believe?

kcgen replied:
Yes; another nice catch @dreamer!
Moved them out of the header

dreamer reviewed (changes requested):
Thank you so much for work on de-templating SoftLimiter and for putting the implementation in cpp file :)

This is review part 1, I believe another rounds of review will be required…

kcgen commented on include/soft_limiter.h:
```diff
@@ -110,7 +110,7 @@ class SoftLimiter {
 	using out_array_t = std::array<int16_t, array_samples>;
 
-	const out_array_t &Apply(const in_array_t &in, uint16_t frames) noexcept;
+	out_array_t Apply(const in_array_t &in, uint16_t frames) noexcept;
```
Fixed - thanks for catching this!

kcgen commented on src/midi/midi_mt32.cpp:
```diff
@@ -440,7 +440,7 @@ void MidiHandler_mt32::Render()
 	while (keep_rendering) {
 		service->renderFloat(buf.data(), FRAMES_PER_BUFFER);
-		const auto &out = soft_limiter.Apply(buf, FRAMES_PER_BUFFER);
+		auto &&out = soft_limiter.Apply(buf, FRAMES_PER_BUFFER);
```
Ahh yes - I now see here:

https://en.cppreference.com/w/cpp/language/copy_elision

that returning a temporary as a plain value, and similarly assigning it as a plain value will let the compiler perform a Return-Value-Optimization on it and simply "place" the result directly into the assigned variable.  So the best of both worlds.. easier to read, and as performant as possible :-)

kcgen commented on include/soft_limiter.h:
```diff
@@ -22,10 +22,18 @@
 #ifndef DOSBOX_SOFT_LIMITER_H
 #define DOSBOX_SOFT_LIMITER_H
 
+#include <algorithm>
+#include <cinttypes>
+#include <limits>
+#include <vector>
+#include <string>
+
+#include "mixer.h"
```
Thanks for catching this!

kcgen commented on include/soft_limiter.h:
```diff
@@ -69,93 +77,79 @@ sequence of audio. For example:
+	            const uint16_t max_frames);
 
-	out_array_t Apply(const in_array_t &in, uint16_t frames) noexcept;
+	using in_t = std::vector<float>;
+	using out_t = std::vector<int16_t>;
```
Done; and does read easier.

kcgen commented on include/soft_limiter.h:
```diff
@@ -69,93 +77,79 @@ sequence of audio. For example:
 private:
-	using in_array_iterator_t = typename std::array<float, array_samples>::const_iterator;
-	using out_array_iterator_t = typename std::array<int16_t, array_samples>::iterator;
+	using in_iterator_t = typename std::vector<float>::const_iterator;
+	using out_iterator_t = typename std::vector<int16_t>::iterator;
```
They're used heavily in many function interfaces; it would get pretty messy with `std::vector<float>::const_iterator` used instead.  Unless there's a better way?

dreamer replied:
I think all ar almost all of functions using these in interfaces should not pass so many parameters list. You can make a simple struct (only in cpp file) and use it to pass input and output from `static` functions (instead of exposing everything through the header file). This would be easier to write, easier to read, easier to understand.

But it's probably out of scope of this PR at this point, so let's leave it be if such refactorization is too much work.

kcgen replied:
yes, the parameter lists are big and I agree a struct to pass would be nicer; but it's it's tricky figuring out how many variables we'd include in the struct -- because the interfaces aren't the same.  

So the struct could be small and cover most use cases (but wouldn't reduce the interface sizes much), or the struct could be mostly all-encompassing, and then some portion of struct members just sit idle (or we pass nullptr's or zeros for them). 

I dunno. Then accessing the values piles on the struct name, so code thats like:

``` c++
while (in_pos != in_end) {
	const auto fitted = poly_a * (*in_pos * prescalar - poly_b) + poly_b;
	assert(fabsf(fitted) <= out_limits::max());
	*out_pos = static_cast<int16_t>(fitted);
	out_pos += 2;
	in_pos += 2;
}
```

becomes:

``` c++
while (vars.in_pos != vars.in_end) {
	const auto fitted = vars.poly_a * (*vars.in_pos * vars.prescalar - vars.poly_b) + vars.poly_b;
	assert(fabsf(fitted) <= out_limits::max());
	*vars.out_pos = static_cast<int16_t>(fitted);
	vars.out_pos += 2;
	vars.in_pos += 2;
}
```

But I fully agree it would collapse the big argument lists - which I would like to get rid of.

I original had all common arguments as members purely to reduce the interface sizes; however they didn't contribute to the end result and really were just temporaries: so I opted to be more struct with the code an really keep them as temporaries.  Maybe that was the wrong move (hmm).

If we can find a way to refactor without these compromises - I can do it now and keep that on a separate commit.  We also have the extensive unit tests that ensure correct operation.

kcgen replied:
I think significantly improve it (without causing a bunch of duplicate code), it will take some hard thinking and likely restructuring.  I don't have the stomach for it - so best we move this to a later time.

With the unit tests proving hard functionality-correctness test, the low LOC count, and it being extremely stand-alone in nature (ie: there's nothing "DOSBox-specific" about the code), all of these lend themselves to any community developers wanting to help!

kcgen commented on include/soft_limiter.h:
```diff
@@ -69,93 +77,79 @@ sequence of audio. For example:
 
 You can then repeatedly call:
-  const auto out_array = limiter.Apply(in_array, num_frames);
+  auto &&out = limiter.Apply(in, num_frames);
```
Thanks for catching these. Fixed.

kcgen commented:
The three new commits address the review feedback thus far (in GUS, SoftLimiter, and MT-32 respectively).

dreamer commented on src/midi/midi_mt32.h:
```diff
@@ -42,19 +42,14 @@
-	using render_buffer_t = std::array<float, SAMPLES_PER_BUFFER>;
-	using limited_buffer_t = std::array<int16_t, SAMPLES_PER_BUFFER>;
-	using ring_t = moodycamel::BlockingReaderWriterCircularBuffer<limited_buffer_t>;
+	using ring_t = moodycamel::BlockingReaderWriterCircularBuffer<std::vector<int16_t>>;
```
Thank you :)

dreamer commented on src/midi/midi_mt32.h:
```diff
@@ -71,18 +66,18 @@ class MidiHandler_mt32 final : public MidiHandler {
 	// Managed objects
 	channel_t channel{nullptr, MIXER_DelChannel};
-	limited_buffer_t buffer{};
+	std::vector<int16_t> play_buffer{};
```
I think this buffer should be initialized to a specific size in `MidiHandler_mt32` c-tor initialization list?

kcgen replied:
The `play_buffer` never sized; it's only ever assigned when dequeue'ing from the ring 
(rqueue `std::move` the content into it, but otherwise this is an assign in our less sophisticated queue too).

The stuff that's assigned into it comes from the SoftLimiter's output (`auto out = limiter... `), which is enqueued into the ring.

The ring only transacts in enqueued objects, so at least one entire rendering pass and enqueue needs to complete, (the ring never hands back and empty or default `T` if it's empty); instead it just waits and blocks.

kcgen replied:
Infact, this is a good question for you:

Is there any way we can instantiate `play_buffer` as a container-for-movement-only? ie: it will only ever be `moved` in.  Something like `std::vector<int16_t> &&play_buffer = {};`?

dreamer replied:
Fortunately it's not possible.

kcgen replied:
> Fortunately it's not possible.

:sweat_smile:

dreamer commented on src/hardware/soft_limiter.cpp:
```diff
@@ -0,0 +1,300 @@
+/*
```
This file does not belong to `hardware` module. No hardware is being emulated here. Please move it to `misc` module.

kcgen replied:
Done! (last commit in round 2)

dreamer reviewed (changes requested):
OK, there's clear improvement. Round 2 of review is done.

kcgen commented on include/soft_limiter.h:
```diff
@@ -60,389 +70,102 @@ needed; typically ranging from 10's of milliseconds to low-hundreds for large >
 public:
-	SoftLimiter(const std::string &name, const AudioFrame &scale);
+	using in_iterator_t = typename std::vector<float>::const_iterator;
+	using out_limits = std::numeric_limits<int16_t>;
+	constexpr static float bounds = static_cast<float>(out_limits::max() - 1);
```
This was a carry over from when the limiter's values were templates as well; but indeed - all of this goes away now :-)

kcgen replied:
`bounds = max - 1`

Checking if a sample's absolute `value` is within `bounds`  keeps readers mind in simple "math" space ... where as comparing again `INT16_...` moves back into programmer type-size space", which is a bit less nice on they eyes.  But yeah.. the values differ by one, so its more strictly correct.

kcgen replied:
`bounds` is used to compare against absolute values, so I tried swapping in `INT16_MAX` in place (also knowing that magnitude-wise it's less than `INT16_MIN` which goes to -32768).  However, what's interesting is that we lose some dynamic range:

``` text
Expected equality of these values:
  out
    Which is: { -4, 32000, 32767, 32000, 2, 32000 }
  expected
    Which is: { -4, 32000, 32766, 32000, 2, 32000 }
[  FAILED  ] SoftLimiter.OutOfBoundsLeftChannel (0 ms)
[ RUN      ] SoftLimiter.OutOfBoundsRightChannel
soft_limiter.cpp:84: Failure
Expected equality of these values:
  out
    Which is: { 32000, -1, 32000, 32767, 32000, 2 }
  expected
    Which is: { 32000, -1, 32000, 32765, 32000, 2 }
[  FAILED  ] SoftLimiter.OutOfBoundsRightChannel (0 ms)
[ RUN      ] SoftLimiter.OutboundsBothChannelsPositive
soft_limiter.cpp:97: Failure
Expected equality of these values:
  out
    Which is: { -4, -1, 32767, 32767, 2, 2 }
  expected
    Which is: { -4, -1, 32766, 32765, 2, 2 }
[  FAILED  ] SoftLimiter.OutboundsBothChannelsPositive (0 ms)
[ RUN      ] SoftLimiter.OutboundsBothChannelsNegative
soft_limiter.cpp:110: Failure
Expected equality of these values:
  out
    Which is: { -4, -1, -32767, -32767, 2, 2 }
  expected
    Which is: { -4, -1, -32766, -32765, 2, 2 }
[  FAILED  ] SoftLimiter.OutboundsBothChannelsNegative (0 ms)
[ RUN      ] SoftLimiter.OutboundsBothChannelsMixed
soft_limiter.cpp:124: Failure
Expected equality of these values:
  out
    Which is: { 20000, -13333, 32767, -32767, 20000, -13333 }
  expected
    Which is: { 19999, -13332, 32766, -32766, 19999, -13332 }
[  FAILED  ] SoftLimiter.OutboundsBothChannelsMixed (0 ms)
[ RUN      ] SoftLimiter.OutboundsBigOneReleaseStep
soft_limiter.cpp:140: Failure
Expected equality of these values:
  out
    Which is: { -17920, 13435 }
  expected
    Which is: { -17920, 13434 }
[  FAILED  ] SoftLimiter.OutboundsBigOneReleaseStep (0 ms)
[ RUN      ] SoftLimiter.OutboundsBig600ReleaseSteps
soft_limiter.cpp:157: Failure
Expected equality of these values:
  out
    Which is: { -32767, 32767 }
  expected
    Which is: { -32766, 32766 }

```

So I'll keep using the limits, but in any case, I've moved it from the interface into the CPP :-)

kcgen commented on src/hardware/gus.cpp:
```diff
@@ -1365,67 +1373,67 @@ void Gus::WriteToRegister()
-	// All the registers below here involve voices
-	if (!voice)
+	// All the registers below operated on the target voice
+	if (!target_voice)
```
I agree. 
In this case, I really wanted to use `voice` in the multiple places where we iterator over `voices` (instead of simply `v`), so I had to move this to accommodate. 

 I also think is a beneficial change, because before this it was simply "voice" and didn't give evidence that this "voice" was actually being selected or targets by the DOS application.  Now it clearer - the DOS application actual targets a specific voice to work on via the register calls.

kcgen commented on src/hardware/gus.cpp:
```diff
@@ -440,18 +440,18 @@ void Voice::GenerateSamples(accumulator_array_t &stream,
+	// Add the samples to the render_buffer, angled in L-R space
+	while (buf < last_buf) {
 		float sample = GetSample(ram);
 		sample *= PopVolScalar(vol_scalars);
 		if (dac_enabled) {
-			*v++ += sample * pan_scalar.left;
-			*v++ += sample * pan_scalar.right;
+			*buf++ += sample * pan_scalar.left;
+			*buf++ += sample * pan_scalar.right;
 		}
```
Yes too all; code is looking much better (and safer) now. 
Surprised none of the static analyzers didn't catch that potential endless loop!

dreamer replied:
Maybe it was not an endless loop? Why was that code hidden behind an if-statement?

dreamer replied:
I made a quick look and don't see any side-effects that might've been tweaking `dac_enabled` on and off. But I don't know GUS code - let's hope this is not a regression.

kcgen commented on src/hardware/gus.cpp:
```diff
@@ -600,6 +600,11 @@ Gus::Gus(uint16_t port, uint8_t dma, uint8_t irq, const std::string &ultradir)
 		voices.at(i) = std::make_unique<Voice>(i, voice_irq);
 	}
 
+	// Size our buffer with the maximum amount of space we'll need
+	// at construction time to avoid any form of allocation during
+	// realtime operation.
+	render_buffer.resize(BUFFER_SAMPLES);
```
Thank for the explanation @dreamer.
Take care of by sizing the vector in the `::GUS(..)` constructor's initializer list.

kcgen commented on include/soft_limiter.h:
```diff
@@ -60,389 +70,102 @@ needed; typically ranging from 10's of milliseconds to low-hundreds for large >
-	using out_array_t = std::array<int16_t, array_samples>;
+	SoftLimiter(const std::string &name,
+	            const AudioFrame &scale,
+	            const uint16_t max_frames);
```
Fixed these and more in the GUS class.

kcgen commented on include/soft_limiter.h:
```diff
@@ -60,389 +70,102 @@ needed; typically ranging from 10's of milliseconds to low-hundreds for large >
-	in_array_iterator_t zero_cross_right = {};
-	in_array_iterator_t precross_peak_pos_left = {};
-	in_array_iterator_t precross_peak_pos_right = {};
+	in_iterator_t zero_cross_left = {};
+	in_iterator_t zero_cross_right = {};
+	in_iterator_t precross_peak_pos_left = {};
+	in_iterator_t precross_peak_pos_right = {};
```
Great catch; these indeed could be moved exactly as suggested!

kcgen commented on src/hardware/gus.cpp:
```diff
@@ -645,22 +650,24 @@ void Gus::AudioCallback(const uint16_t requested_frames)
 		}
-		const auto &out_stream = soft_limiter.Apply(accumulator, frames);
-		audio_channel->AddSamples_s16(frames, out_stream.data());
+		const auto &&out = soft_limiter.Apply(render_buffer, frames);
```
Ahh thanks! Got it.

kcgen commented:
Thanks for the extremely thorough round 2 @dreamer! The code is getting very nice now.
I've put all the round-2 changes on the last three commits. 
Regarding final squashing - it might make sense to keep all the review-improvements separate from the original "type-change-only" commits. That way if we do hit a regression at some point, we can atleast differentiate if it was the vector-type change or this other cleanup.

dreamer commented on vs/dosbox.vcxproj.filters:
```diff
@@ -367,6 +367,9 @@
       <Filter>src\misc</Filter>
     </ClCompile>
+    <ClCompile Include="..\src\misc\soft_limiter.cpp">
+      <Filter>src\hardware</Filter>
```
Misleading filter.

dreamer commented on src/misc/soft_limiter.cpp:
```diff
@@ -0,0 +1,309 @@
+{}
+
+//  Limit the input array and returned as integer array
+typename std::vector<int16_t> SoftLimiter::Process(const std::vector<float> &in,
```
This `typename` is unnecessary.

kcgen replied:
Thank you!

kcgen commented:
Last couple commits: 

 - SoftLimiter's stat calculation assumed each pass is 1ms, but that's not true for FluidSynth and MT-32 (they operate on longer sequences). During testing I noticed it wasn't working right for FluidSynth.  So this fixes that.
 - With it working nicely now, I also added stats printing for MT-32, which I forgot to do when we switched to using mt32emu's floating-pointer renderer.

Testing (and stats printing) now working correctly for all three floating-point audio devices.

I think this PR is pretty wrung out!  Shout if you see anything else :-)

dreamer commented on src/midi/midi_fluidsynth.cpp:
```diff
@@ -368,19 +373,20 @@ void MidiHandlerFluidsynth::PrintStats()
+
+	while (requested_frames > 0) {
+		const uint16_t frames = std::min(requested_frames, max_frames);
+		assert(frames <= max_frames);
```
I think this assert is an overkill :) It's calculated literally a line above…

kcgen replied:
Agreed; knocking it out :-)

dreamer reviewed (changes requested):
OK, last few comments - trivial stuff :)

Please either squash or properly describe the commits - as you see fit.

kcgen commented:
Squashed and re-tested.
Thanks again @dreamer - your review took this code up several notches!

kcgen commented:
Just a side comment - I'm not sure why, but GitHub is allowing me to squash and merge this PR the base branch, even though the PR hasn't been approved.  

![2021-02-17_21-36](https://user-images.githubusercontent.com/1557255/108310625-88c3c180-7168-11eb-8838-6feccad7ce07.png)

I've be able to do some trivial merges back in January one they passed CI; but I can't now (such as cherry-picking and backporting the CD-DA pregap start fix to 0.76.x). 

Are you able to see anything on your side?

dreamer reviewed (approved):
OK, looks great - @kcgen thanks :)
